[WhyAre] iP

WhyAre opened the pull request:
# Heading

- bullet list

1. numbered list

```java
System.out.println("I use vim btw");
```

- [ ] task list

:sob: 

> block quote

[Click here to install neovim](https://neovim.io)

`inline code`

I use **vim** (~instead of emacs~) _btw_

WhyAre committed:
Rename all occurrences of Duke to JankyBot

WhyAre committed:
Add feature to greet user and exit

WhyAre committed:
Extract greet and bye into functions

WhyAre committed:
Add feature to echo user input

WhyAre committed:
Add feature to store and retrieve user text

WhyAre committed:
Add feature to mark as done

WhyAre committed:
Refactor code to not use switch statements

WhyAre committed:
Refactor command parsing to handle string array instead of string

WhyAre committed:
Add todo, deadline and event tasks

WhyAre committed:
Rename variable from memory to tasks

WhyAre committed:
Add regression test

WhyAre committed:
Add error handling

WhyAre committed:
Implement delete feature

WhyAre committed:
Change name to JankBot

WhyAre committed:
Fix test cases

WhyAre committed:
Use Collectors.joining to express intent more clearly

WhyAre committed:
Rename variable from c to i

Any sane person would name their counter i

WhyAre committed:
Change other occurrences of JankyBot to JankBot

WhyAre committed:
Add data folder to gitignore

WhyAre committed:
Make task serializable

WhyAre committed:
Add save functionality

WhyAre committed:
Merge branch 'branch-Level-7'

WhyAre committed:
Store date as LocalDateTime instead of String

WhyAre committed:
Merge branch 'branch-Level-8'

WhyAre committed:
Add TaskList class

WhyAre committed:
Add print function to TaskList class

WhyAre committed:
Add Storage class

WhyAre committed:
Make Storage work with TaskList instead of ArrayList<Task>

WhyAre committed:
Make TaskList serializable

WhyAre committed:
Make method return Task instead of void

WhyAre committed:
Add remove method to TaskList

WhyAre committed:
Add size method to TaskList

WhyAre committed:
Print something even though there are no tasks

WhyAre committed:
Refactor main class to use newly created classes

WhyAre committed:
Add parser and ui classes

WhyAre committed:
Use more OOP

WhyAre committed:
Add gradle support

WhyAre committed:
Update build.gradle

WhyAre committed:
Unignore .idea folder

Ref: https://intellij-support.jetbrains.com/hc/en-us/articles/206544839-How-to-manage-projects-under-Version-Control-Systems

WhyAre committed:
Save run config from intellij

WhyAre committed:
Move file into package

WhyAre committed:
Update intelliJ config

WhyAre committed:
Change to use List instead of ArrayList

Rely on an interface rather than a class (Dependency Inversion Principle)

WhyAre committed:
Add unit tests

Can add more in the future

WhyAre committed:
Refactor code to conform to a coding style

Additionally, implement checkstyle to programmatically detect violations

WhyAre committed:
Call isEmpty() instead of checking size

WhyAre committed:
Add find feature

WhyAre committed:
Merge branch 'branch-A-CodingStandard'

WhyAre committed:
Merge branch 'branch-Level-9'

WhyAre committed:
Add javadocs to Storage and Task class

WhyAre committed:
Add javadoc for JankBot class

WhyAre committed:
Merge branch 'branch-A-JavaDoc'

WhyAre committed:
Refactor parsing logic

commit b055a505f7190874c8d1b60c89fee8635b34a027
Author: WhyAre <laiyr52@gmail.com>
Date:   Sat Jan 25 16:04:37 2025 +0800

    Add unit tests for parsing

commit ef3f88c9afcd85b63630005c24d67d9fad58f7d4
Author: WhyAre <laiyr52@gmail.com>
Date:   Sat Jan 25 15:53:41 2025 +0800

    Change to use comma instead of newline as delimiter for error message in missing keys

commit 9b98f40b85079df570ff214ccd977a7d88f9261f
Author: WhyAre <laiyr52@gmail.com>
Date:   Sat Jan 25 15:39:16 2025 +0800

    Change input date format

commit 360236cc1e993e382d8da4d1c53c4ca352698a4d
Author: WhyAre <laiyr52@gmail.com>
Date:   Sat Jan 25 15:36:19 2025 +0800

    Move parsing logic into Parser.java

commit 7f73b475dfb173c25b6f15257411419b7cc63492
Author: WhyAre <laiyr52@gmail.com>
Date:   Sat Jan 25 14:50:41 2025 +0800

    Remove curly bracket for single statement in switch

commit 0b3473a256d820937178ebd4fc191cda620cb75f
Author: WhyAre <laiyr52@gmail.com>
Date:   Sat Jan 25 14:50:13 2025 +0800

    Make Task superclass non-abstract and remove parse method

deborahluyx commented on src/main/java/jank/JankBot.java:
```diff
@@ -0,0 +1,140 @@
+     * @param line command as a String[]
+     * @throws JankBotException
+     */
+    static void processCommand(String[] line) throws JankBotException {
```
Should this function name be renamed to provide more explanation on how the command is processed? Perhaps `executeUserCommand`?

deborahluyx commented on src/main/java/jank/Storage.java:
```diff
@@ -0,0 +1,51 @@
+     */
+    static TaskList loadTasks(String filepath) {
+        try (var fin = new FileInputStream(filepath); var ois = new ObjectInputStream(fin)) {
+            if (ois.readObject() instanceof TaskList tasklist) {
```
Perhaps use camel case for this?

deborahluyx reviewed (commented):
Looks really good overall! Not much comments left because the code is good 👍🏻 Keep it up!

WhyAre committed:
Add javafx to project

WhyAre committed:
Add output text area

WhyAre committed:
Change name back to JankBot

WhyAre committed:
Checkstyle to support lambda style switch case

WhyAre committed:
Update UI

WhyAre committed:
Add UI Behavior

WhyAre committed:
Refactor processCommand to return String

WhyAre committed:
Rename processCommand method name

WhyAre committed:
Merge branch 'branch-Level-10'

WhyAre committed:
Fix checkstyle violations

hzk-lab commented on src/main/java/jank/TaskImpl.java:
```diff
@@ -0,0 +1,74 @@
+        var map = Arrays.stream(flags)
+                        .collect(Collectors.toMap(Function.identity(), x -> new StringBuffer(), (left, right) -> right,
+                                HashMap<String, StringBuffer>::new));
+        map.put("", new StringBuffer());
```
I think you may consider adding a brief comment explaining the purpose of map.put("", new StringBuffer());. Right now, it's not immediately clear why an empty key is needed.

hzk-lab commented on src/main/java/jank/Parser.java:
```diff
@@ -0,0 +1,102 @@
+package jank;
```
That's good use of Java Stream here!

hzk-lab commented on src/main/java/jank/TaskList.java:
```diff
@@ -0,0 +1,71 @@
+        return task;
+    }
+
+    Task remove(int index) {
```
I think you can consider handling IndexOutOfBoundsException in remove(int index) to prevent unexpected crashes if the index is invalid.

hzk-lab commented on src/main/java/jank/TaskList.java:
```diff
@@ -0,0 +1,71 @@
+     * @param query Query string
+     * @return list of tasks
+     */
+    List<Task> find(String query) {
```
What happens if query is null? Maybe add a null check before filtering to prevent NullPointerException.

hzk-lab commented on src/main/java/jank/Parser.java:
```diff
@@ -0,0 +1,102 @@
+        boolean isMarked = line[0].equalsIgnoreCase("mark");
+        Parser.checkHasArgs(line, "Usage: %s <index>".formatted((isMarked) ? "mark" : "unmark"));
+
+        int index;
```
It seems that you introduce many try-catch bolcks to read the next line. Why not consider packing this logic in a separate method to avoid repetition?

WhyAre committed:
Beautify user interface

WhyAre committed:
Add feature to exit program by typing bye

WhyAre committed:
Make constants uppercase

WhyAre committed:
Convert static methods in JankBot to non-static

WhyAre committed:
Remove commented code

WhyAre committed:
Made some methods non-static

WhyAre committed:
Create package jank.ui

WhyAre committed:
Delete unused class

WhyAre committed:
Move commands into its own package

WhyAre committed:
Refactor task into its own package

WhyAre committed:
Fix checkstyle issues

WhyAre committed:
Add check to ensure that index provided is within range

WhyAre committed:
Merge branch 'branch-A-CodeQuality'

WhyAre committed:
Add assertions

WhyAre committed:
Add assertion statements

There were no assertions.

Assertions help to minimize programmer mistakes.

Let's add assertion statements throughout the code

WhyAre committed:
Fix tests to work with new package layout

Unit tests does not compile and run properly

Tests are needed to work properly in order to know that the software
works to some extent

Let's fix the unit tests to make sure they compile and run properly

JUnit is used to run the tests

WhyAre committed:
Change formatting rules

WhyAre committed:
Run ./gradlew check in github action

WhyAre committed:
Fix broken workflow rule; branch name should be master not main

WhyAre committed:
Add checkstyle and unit tests to GitHub Actions (#1)

Checkstyle and unit tests are not being run before pushing code to GitHub

These gradle tasks are often forgotten to be executed by the developer before pushing changes

Let's add a github action to run checkstyle and tests on pull request to master

GitHub actions are used so that there isn't a need for additional dependencies on the developer's machine. `./gradlew check` is being executed because it runs both the unit tests and checkstyle

Ref: https://stackoverflow.com/questions/50104666/difference-between-test-and-check-tasks-in-gradle

WhyAre committed:
Add remind command (#2)

There is no remind command

Remind command is to help the users print out the deadlines within an interval (e.g. next 5 days)

Let's add the remind command functionality to the chatbot

The command syntax is `remind 5d` which means to show reminders in the next 5 days.
It works with `d`, `h`, `m`, `s` (days, hours, minutes, seconds respectively)

WhyAre committed:
Add sorting functionality (#3)

There is no sort command

Sort command is convenient to show upcoming deadlines

Let's add a sort command.

The sort command places `TodoEvents` at the bottom, `DeadlineTask` is sorted based off of `/by` date, `EventTask` is sorted using the `/to` date.

WhyAre committed:
Set java version to 17

WhyAre committed:
Add launcher class

WhyAre committed:
Make each input-output pair be "grouped" in the user interface

WhyAre committed:
Change welcome message

WhyAre committed:
Add brief user guide (#5)

WhyAre committed:
Add workflow to create GitHub release (#6)

* Add release workflow

* Rename workflow name

WhyAre committed:
Fix checkstyle complaints

WhyAre committed:
Add a better GUI (#8)

There is no separator between user commands

Let's add a horizontal separator between commands

This is so that the user can easily distinguish the output of multiple commands

WhyAre committed:
Rename data file to bin (#9)

It saves as a file with `.txt` extension

This is inaccurate because it's a binary file

Let's change the extension to `.bin` instead

WhyAre committed:
Fix issue of jar not being cross-platform

Jar isn't cross-platform, it can work on either windows/linux and not both

It should work cross-platform

Jar is made cross platform by adding all the dependencies that it would ever need

Ref: https://stackoverflow.com/questions/64756006/not-able-to-generate-cross-platform-javafx-application

WhyAre committed:
Change error handling if serialized version doesn't match (#11)

Used to throw error if the version of the serialized data doesn't match

The program will crash if the code is updated

Let's make the program not crash when the codebase is updated

One caveat is that the code will "fail silently" if the config file is corrupted for any reason.
The user will not know but the bot will silently recover from it

WhyAre committed:
Add smoke test in GitHub Action
